### PR TITLE
Stacked contexts

### PR DESCRIPTION
**Pull request by prescod**
My goal is to redo the faker stuff as a "built-in plug-in".

David Glick had suggested a context in which the faker-feature might want to have "contextual state".

So that got me thinking about plugins with contextual state.

But before I implemented that, I wanted to clean up the RuntimeContext class which had accumulated a ton of different features and state. So I refactored it so that it could have just two purposes:

1. Be the holder for any "stack-context" for Snowfakery processes.

2. Be a Facade for the rest of the interpreter in most contexts.

The Interpreter itself is a new object that holds the global state of the process (options, plugins, etc.) as opposed to the local state which stays in the RuntimeContext.

The interpreterholds a pointer to the "current context" (RuntimeContext) so that Plugins can have a persistent handle to the interpreter and grab transient references to the current runtime context.

In addition to factoring the Interpreter out of the RuntimeContext, I also factored out the FinishedChecker which keeps track of whether the interpreter is finished iterating over the Snowfakery script.

This is all still groundwork for reworking some of the built-in facilities, like Faker and template_funcs into built-in plug-ins.

### CODE REVIEW

File: snowfakery/data_generator_runtime.py [hunk 1]
```diff
@@ -130,126 +183,140 @@ def get_evaluator(self, definition: str):
...
-        )
+        "Create a nested RuntimeContext (analogous to a 'stack frame')."
+        jr = self.__class__(current_table_name=tablename, interpreter=self.interpreter, parent_context=self)
+        # jr will register itself with the interpreter
+        try:
+            yield jr
+        finally:
+            # Goodbye junior, its been nice
+            # Hope you find your paradise
+            self.interpreter.current_context = self
```
[the remark this answers is not in the document]
Comment (prescod) [reply snowfakery/data_generator_runtime.py:267]: I don't think that the parent ever gets a pointer to the junior, so there's no cycle.

Junior still has two pointers "up" (to the interpreter and the parent context) but I don't think there are any pointers down.

[the remark this answers is not in the document]
Comment (davisagli) [reply snowfakery/data_generator_runtime.py:267]: Ah right, not sure what I was thinking.

File: snowfakery/data_generator_runtime.py [hunk 2]
```diff
@@ -258,21 +325,25 @@ def curry(func):
...
 
 class DynamicEvaluator:
     def __init__(self, template):
         self.template = template
 
     def __call__(self, context):
-        return self.template.render(**context.field_vars(), **context.field_funcs())
+        namelookup = context.evaluation_namespace
+        return self.template.render(
+            {**namelookup.field_vars(), **namelookup.field_funcs()}
```
[the remark this answers is not in the document]
Comment (prescod) [reply snowfakery/data_generator_runtime.py]: That inspired me to eliminate this whole class. Once you remove the merging its basically just a glorified curry and not the delicious kind.

80e10e343d8

https://www.youtube.com/watch?v=o9pEzgHorH0

File: snowfakery/plugins.py [hunk 3]
```diff
@@ -19,17 +16,48 @@ def func1(self, arg1, arg2, arg3=default):
...
+
+class PluginContext:
+    "Exposes certain stable internals to plugins"
+
+    def __init__(self, plugin):
+        self.plugin = plugin
+        self.interpreter = plugin.interpreter
+
+    def field_vars(self):
+        return self.interpreter.current_context.field_vars()
```
Comment (davisagli) [inline snowfakery/plugins.py:60]: Is it desired or not that this now includes the field_funcs?

Comment (prescod) [reply snowfakery/plugins.py:60]: Yes I think its better to be consistent. Now plugins can use the field_funcs in case they are useful.

Comment (prescod) [reply snowfakery/plugins.py:60]: I think that the only reason they were ever separate is because I had to do some curry magic with the field_funcs.

**Comment by prescod** [comment]
Merged and I'll fix the black stuff in the (an) other commit.

**Pull request merged**